lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+
+    Args:
+        service: The name of the service which has changed state.
+        state: The state of the service.
+    """
+    if service.endswith(".service"):
+        service = service[0 : -len(".service")]
+    if state == "active":
+        event_name = "started"
+    else:
+        event_name = "stopped"
+    hook = f"service-{service}-{event_name}"
+    cmd = ["/usr/bin/juju-exec", _juju_unit, f"hooks/{hook}"]
+
+    _logger.debug(f"Invoking hook {hook} with command: {' '.join(cmd)}")
+    process = await asyncio.create_subprocess_exec(
+        *cmd,
+    )
+    await process.wait()
+    if process.returncode:
+        _logger.error(
+            f"Hook command '{' '.join(cmd)}' failed with returncode {process.returncode}"
+        )
+    else:
+        _logger.info(f"Hook command '{' '.join(cmd)}' succeeded.")
+
+
+async def _get_state(bus: MessageBus, service: str) -> str:
+    """Report the current state of a service.
+
+    Args:
+        bus: The message bus to query on.
+        service: The systemd service to query the state of.
+
+    Returns:
+        The state of the service. "active" or "inactive"
+    """
+    obj_path = _name_to_dbus_path(service)
+    try:
+        _logger.debug(f"Retrieving state for service {service} at object path: {obj_path}")
+        introspection = await bus.introspect("org.freedesktop.systemd1", obj_path)
+        proxy = bus.get_proxy_object("org.freedesktop.systemd1", obj_path, introspection)
+        properties = proxy.get_interface("org.freedesktop.DBus.Properties")
+        state = await properties.call_get("org.freedesktop.systemd1.Unit", "ActiveState")  # noqa
+        return state.value
+    except DBusError:
+        # This will be thrown if the unit specified does not currently exist,
+        # which happens if the application needs to install the service, etc.
+        return "unknown"
+
+
+async def _async_load_services() -> None:
+    """Load names of services to observe from legacy Juju hooks.
+
+    Parses the hook names found in the charm hooks directory and determines
+    if this is one of the services that the charm is interested in observing.
+    The hooks will match one of the following names:
+
+      - service-{service_name}-started
+      - service-{service_name}-stopped
+
+    Any other hooks are ignored and not loaded into the set of services
+    that should be watched. Upon finding a service hook it's current ActiveState
+    will be queried from systemd to determine it's initial state.
+    """
+    global _juju_unit
+    hooks_dir = Path.cwd() / "hooks"
+    _logger.info(f"Loading services from hooks in {hooks_dir}")
+
+    if not hooks_dir.exists():
+        _logger.warning(f"Hooks dir {hooks_dir} does not exist.")
+        return
+
+    watched_services = []
+    # Get service-{service}-(started|stopped) hooks defined by the charm.
+    for hook in filter(lambda p: _service_hook_regex_filter.match(p.name), hooks_dir.iterdir()):
```

Comment:
We're doing the `match` call twice. Besides, I think this would be simpler and clearer as:

```python
for hook in hooks_dir.iterdir():
    match = _service_hook_regex_filter.match(hook.name)
    if match:
        watched_services.append(match.group("service"))
```

Or -- I'm not the biggest fan of the walrus operator (Python 3.8+), but it does work pretty well here:

```python
watched = [match.group("service") for h in hooks_dir.iterdir()
           if (match := _service_hook_regex_filter.match(h.name))]
```